timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/BlockingTimeouts.java

```diff
@@ -0,0 +1,96 @@
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.atlasdb.timelock.config.TimeLockServerConfiguration;
+
+import io.dropwizard.util.Duration;
+
+public final class BlockingTimeouts {
+    private static final Logger log = LoggerFactory.getLogger(BlockingTimeouts.class);
+
+    public static final long DEFAULT_IDLE_TIMEOUT = 30 * 1000;
+
+    private static final double ERROR_MARGIN = 0.03;
+
+    private static final String SERVER = "server";
+    private static final String APPLICATION_CONNECTORS = "applicationConnectors";
+    private static final String IDLE_TIMEOUT = "idleTimeout";
+
+    private BlockingTimeouts() {
+        // utility
+    }
+
+    public static long getBlockingTimeout(ObjectMapper mapper, TimeLockServerConfiguration configuration) {
+        // TODO (jkong): Need to decide if this is worth the pain to enforce our invariant, or if we should just
+        // let users do it manually!
+        if (mapper == null || configuration == null) {
+            log.warn("Attempted to initiate TimeLock with a missing object mapper or null configuration."
+                    + " Using the default timeout of {} ms", DEFAULT_IDLE_TIMEOUT);
+            return getDefaultBlockingTimeout();
+        }
+        try {
+            List<Map<String, String>> connectorData = getConnectorData(mapper, configuration);
+            long minimumTimeout = getMinimumTimeout(connectorData);
+            return scaleForErrorMargin(minimumTimeout);
+        } catch (IOException e) {
+            log.warn("Couldn't figure out the idle timeout from configuration. Will assume this to be the"
+                    + " default of {} ms", DEFAULT_IDLE_TIMEOUT);
+            return getDefaultBlockingTimeout();
+        }
+    }
+
+    public static long getDefaultBlockingTimeout() {
+        return scaleForErrorMargin(DEFAULT_IDLE_TIMEOUT);
+    }
+
+    @SuppressWarnings("unchecked") // Any map can be deserialized as Map<String, String>.
+    private static List<Map<String, String>> getConnectorData(ObjectMapper mapper,
+            TimeLockServerConfiguration configuration) throws JsonProcessingException {
+        JsonNode configurationJson = mapper.valueToTree(configuration);
+        JsonNode connectorJson = configurationJson.get(SERVER).get(APPLICATION_CONNECTORS);
+        return Arrays.asList(mapper.treeToValue(connectorJson, Map[].class));
+    }
+
+    private static long getMinimumTimeout(List<Map<String, String>> connectorData) {
+        return connectorData.stream()
+                .map(connectorDatum -> connectorDatum.get(IDLE_TIMEOUT))
+                .map(idleTimeout -> Duration.parse(idleTimeout).toMilliseconds())
+                .min(Long::compareTo)
+                .orElse(DEFAULT_IDLE_TIMEOUT);
```

Comment:
Hmm, yep. This would mean there were no connectors in the serialized JSON which seems impossible (I _think_ - since DW will have initialised these when starting the server - but probably worth checking). Will separate the "unknown" logging to its own method and have that do the work.